parachain/pallets/basic-channel/src/outbound/mod.rs

```diff
@@ -215,56 +269,121 @@ pub mod pallet {
 			Ok(())
 		}
 
+		/// For every account id in the enqueued messages, create an Eth ABI-encoded message
+		/// bundle containing the messages for that account. Hash the ethabi encoding of these
+		/// message bundles with a Merkle tree to produce a commitment hash. Then store the
+		/// commitment on the parachain, emit an event with the commitment and SCALE-encoded
+		/// message bundles and persist the ethabi-encoded message bundles to off-chain storage.
 		fn commit() -> Weight {
-			let messages = <MessageQueue<T>>::take();
-			if messages.is_empty() {
+			// TODO: consider using mutate_exists here. If some part of emitting message bundles
+			// fails, we don't want the MessageQueue to be empty.
+			let message_queue = <MessageQueue<T>>::take();
+			if message_queue.is_empty() {
 				return T::WeightInfo::on_initialize_no_messages();
 			}
 
-			let nonce = <Nonce<T>>::get();
-			let next_nonce = nonce.saturating_add(1);
-			<Nonce<T>>::put(next_nonce);
+			// Store these for the on_initialize call at the end
+			let message_count = message_queue.len() as u32;
+			let average_payload_size = Self::average_payload_size(&message_queue);
 
-			let bundle = MessageBundle {
-				source_channel_id: ChannelId::Basic as u8,
-				nonce: next_nonce,
-				messages: messages.clone(),
-			};
+			let message_bundles = Self::make_message_bundles(message_queue);
+			let eth_message_bundles: Vec<Vec<u8>> = message_bundles
+				.clone()
+				.into_iter()
+				.map(|bundle| ethabi::encode(&vec![bundle.into()]))
+				.collect();
+
+			// TODO: Maybe undo the O type param in merkle_root, since we can convert between H256
```

Comment:
Yes please. Internally in `merkle_proof.rs` Its fine to hardcode `[u8; 32]` as the hash type and remove the `O` parameter. It will never change.